MCprep_addon/spawner/mcmodel.py

```diff
@@ -0,0 +1,256 @@
+
+    return mat
+
+def locate_image(textures,img,model_filepath):
+
+    context=bpy.context
+    resource_folder = bpy.path.abspath(context.scene.mcprep_texturepack_path)
+
+    local_path = textures[img]
+    if local_path[0] == '#': #reference to another texture
+        return locate_image(textures,local_path[1:],model_filepath)
+    else:
+        if local_path[0] == '.': #path is local to the model file
+            directory = os.path.dirname(model_filepath)
+        else:
+            if(len(local_path.split(":"))==1):
+                namespace="minecraft"
+            else:
+                namespace=local_path.split(":")[0]
+                local_path=local_path.split(":")[1]
+
+            directory = os.path.join(resource_folder,"assets",namespace,"textures")
+        return os.path.realpath(os.path.join(directory,local_path)+".png")
+
+def read_model(model_filepath): # recursive function that extracts the information from the json files
+    with open(model_filepath, 'r') as f:
+            obj_data=json.loads(f.read())
+    
+    context=bpy.context
+    resource_folder = bpy.path.abspath(context.scene.mcprep_texturepack_path)
+    
+    elements=None
+    textures=None
+
+    parent = obj_data.get("parent")
+    if parent is not None:
+        if parent=="builtin/generated" or parent=="item/generated":
+            pass # generates the model from the texture
+        elif parent=="builtin/entity":
+            pass # model from an entity file, only for chests, ender chests, mob heads, shields, banners and tridents
+        else:
+            if(len(parent.split(":"))==1):
+                namespace="minecraft"
+                parent_filepath=parent
+            else:
+                namespace=parent.split(":")[0]
+                parent_filepath=parent.split(":")[1]
+            
+            models_dir = os.path.join(resource_folder,"assets",namespace,"models")
+            elements, textures = read_model(os.path.join(models_dir,parent_filepath)+".json")
+    
+    current_elements=obj_data.get("elements")
+    if current_elements is not None:
+        elements = current_elements # overwrites any elements from parents
+
+    current_textures=obj_data.get("textures")
+    if current_textures is not None:
+        if textures is None: textures=current_textures
+        else:
+            for img in current_textures:
+                textures[img]=current_textures[img]
+
+    conf.log("\nfile:"+str(model_filepath))
+    conf.log("parent:"+str(parent))
+    conf.log("elements:"+str(elements))
+    conf.log("textures:"+str(textures))
+
+    return elements, textures
+
+def add_model(model_filepath, obj_name="MinecraftModel"):
+    
+    mesh = bpy.data.meshes.new(obj_name)  # add a new mesh
+    obj = bpy.data.objects.new(obj_name, mesh)  # add a new object using the mesh
+
+    collection = bpy.context.collection
+    view_layer = bpy.context.view_layer
+    collection.objects.link(obj)  # put the object into the scene (link)
```

Comment:
FYI only: I will need to do some extra work after the fact to make this cross compatible with Blender 2.7x (using groups instead of collections), but I won't make you worry about this in your pull request.